mcsema/BC/Function.cpp

```diff
@@ -260,6 +289,132 @@ static llvm::BasicBlock *GetOrCreateBlock(TranslationContext &ctx,
   return block;
 }
 
+static void LiftExceptionFrameLP(TranslationContext &ctx,
+                                 const NativeFunction *cfg_func) {
+
+  if (cfg_func->eh_frame.size() > 0) {
+
+    // The personality function `__gxx_personality_v0` is lifted as global variable. Erase the
+    // variable before declaring it as the function.
+    if (auto personality_func_var = gModule->getGlobalVariable("__gxx_personality_v0")) {
+      personality_func_var->eraseFromParent();
+    }
+
+    auto personality = gModule->getFunction("__gxx_personality_v0");
+
+    if (personality == nullptr) {
+      personality = llvm::Function::Create(
+          llvm::FunctionType::get(llvm::Type::getInt32Ty(*gContext), true),
+          llvm::Function::ExternalLinkage, "__gxx_personality_v0", gModule);
+    }
+
+    auto lifted_func = gModule->getFunction(cfg_func->lifted_name);
+    lifted_func->addFnAttr(llvm::Attribute::UWTable);
+    lifted_func->addFnAttr(llvm::Attribute::OptimizeNone);
+    lifted_func->removeFnAttr(llvm::Attribute::NoUnwind);
+#if LLVM_VERSION_NUMBER > LLVM_VERSION(3, 6)
+    lifted_func->setPersonalityFn(personality);
+#endif
+  }
+
+  for (auto &entry : cfg_func->eh_frame) {
+    if (entry->lp_ea == 0) continue;
+
+    if (entry->action == 0) {
+      std::stringstream ss;
+      ss << "cleanup_landingpad_" << std::hex << entry->lp_ea;
+      llvm::BasicBlock *landing_bb = llvm::BasicBlock::Create(
+          *gContext, ss.str(), ctx.lifted_func);
+      llvm::IRBuilder<> ir(landing_bb);
+      auto exn_type = llvm::StructType::get(llvm::Type::getInt8PtrTy(*gContext),
+                                            llvm::Type::getInt32Ty(*gContext), nullptr);
+#if LLVM_VERSION_NUMBER > LLVM_VERSION(3, 6)
+      auto lpad = ir.CreateLandingPad(exn_type, 1, "cleanup.lpad");
+#else
+      auto personality = gModule->getFunction("__gxx_personality_v0");
+      auto lpad = ir.CreateLandingPad(exn_type, personality, 1, "cleanup.lpad");
+#endif
+      lpad->setCleanup(true);
+
+      auto exctn_0  = ir.CreateExtractValue(lpad, 0);
+      auto store_loc_0 = ir.CreateAlloca(exctn_0->getType());
+      ir.CreateStore(exctn_0, store_loc_0);
+
+      auto exctn_1 = ir.CreateExtractValue(lpad, 1);
+      auto store_loc_1 = ir.CreateAlloca(exctn_1->getType());
+      ir.CreateStore(exctn_1, store_loc_1);
+
+      auto lp_entry = ctx.ea_to_block[entry->lp_ea];
+     // landing_bb->dump();
+      if (nullptr != lp_entry) {
```

Comment:
I think you should be using `GetOrCreateBlock` on the `entry->lp_ea`, or document valid cases where an entry ought not be associated with a block. To me this even says that maybe we're not lifting the whole function, and that on the CFG production side, we should be extending the function with any reachable `lp_ea`s.

```c++
LOG_IF(ERROR, !ctx.ea_to_block.count(entry->lp_ea))
    << "Missing block at " << std::hex << entry->lp_ea
    << " for exception landing pad from " << TODO << std::dec;
auto lp_entry = GetOrCreateBlock(ctx, entry->lp_ea);
ir.CreateBr(lp_entry);
ctx.lp_to_block[entry->lp_ea] = landing_bb;
```

Reply:
It's possible that I am not aware of some behaviour of exceptions, e.g. if the `lp_ea` is zero then it means we don't catch it or something. If that's the case then you need to document this behaviour in the code with a comment.